src/mm/dbcsr_mm_cannon.F

```diff
@@ -1329,6 +1361,73 @@ SUBROUTINE multiply_cannon(left_set, right_set, product_matrix, &
          trs_stackbuf_calc => trs_stackbuf_1
          trs_stackbuf_comm => trs_stackbuf_2
       END IF
+
+      ! Initial upload for acc_centric version
+      ! We now send the transposed GPU buffer instead of the original buffer
+      IF (acc_centric) THEN
+
+#ifdef __parallel
+         CALL mp_environ(numtask=mpi_size, groupid=mp_group)
+         IF (mpi_size > 1 .AND. .NOT. mp_query_hip_support() .AND. .NOT. mp_query_cuda_support()) THEN
+            DBCSR_ABORT("acc_centric mode requires a CUDA/HIP-aware installation of MPI.")
+         END IF
+#endif
+
+         CALL timeset(routineN//"_acc_init", handle1)
+         DO v_ki_right = 1, right_row_nimages
+            CALL dbcsr_data_host2dev(right_buffer_calc%mats(v_ki_right, 1)%data_area)
+         END DO
+         DO v_ki_left = 1, left_col_nimages
+            CALL dbcsr_data_host2dev(left_buffer_calc%mats(1, v_ki_left)%data_area)
+         END DO
+
+         v_ki_left = 1
+         v_ki_right = 1
+         DO v_ki = 1, max(left_col_nimages, right_row_nimages)
+            IF (v_ki_right .GT. right_row_nimages) THEN
+               DBCSR_ABORT("Failed assumption left_col_nimages >= right_row_nimages")
```

Comment:
In my tests, I was able to trigger this failure. I am not sure how many real-world use cases this assumption excludes from `acc_centric` mode, might want to consider the opposite case, too.